rtl/eclrtl/eclregex.cpp

```diff
@@ -121,82 +182,127 @@ class CStrRegExprFindInstance : implements IStrRegExprFindInstance
 class CCompiledStrRegExpr : implements ICompiledStrRegExpr
 {
 private:
-    regex    regEx;
+    pcre2_code_8 * compiledRegex = nullptr;
+    pcre2_match_data_8 * matchData = nullptr;
 
 public:
-    CCompiledStrRegExpr(const char * _regExp, bool _isCaseSensitive = false)
+    CCompiledStrRegExpr(const char * _regex, bool _isCaseSensitive = false)
     {
-        try
-        {
-            if (_isCaseSensitive)
-                regEx.assign(_regExp, regex::perl);
-            else
-                regEx.assign(_regExp, regex::perl | regex::icase);
-        }
-        catch(const boost::bad_expression & e)
+        int errNum = 0;
+        PCRE2_SIZE errOffset;
+        uint32_t options = (_isCaseSensitive ? 0 : PCRE2_CASELESS);
+
+        compiledRegex = pcre2_compile_8((PCRE2_SPTR8)_regex, PCRE2_ZERO_TERMINATED, options, &errNum, &errOffset, pcre2CompileContext8);
+
+        if (compiledRegex == nullptr)
         {
-            std::string msg = "Bad regular expression: ";
-            msg += e.what();
-            msg += ": ";
-            msg += _regExp;
-            rtlFail(0, msg.c_str());  //throws
+            failWithPCRE2Error(errNum, "Error in regex pattern: ", _regex);
         }
+
+        matchData = pcre2_match_data_create_from_pattern_8(compiledRegex, pcre2GeneralContext8);
+    }
+
+    ~CCompiledStrRegExpr() //CAVEAT non-virtual destructor !
+    {
+        pcre2_match_data_free_8(matchData);
+        pcre2_code_free_8(compiledRegex);
     }
 
     //ICompiledStrRegExpr
 
     void replace(size32_t & outlen, char * & out, size32_t slen, char const * str, size32_t rlen, char const * replace) const
     {
-        std::string src(str, str + slen);
-        std::string fmt(replace, replace + rlen);
-        std::string tgt;
-        try
+        PCRE2_SIZE pcreLen = 0;
```

Comment:
This probably needs protecting with a critical section - in the rare case that multiple threads are performing a regex replace on using the same pattern at the same time.
Alternatively matchData could be allocated within this call.  That might be a better approach since it would avoid the allocation when only REGEXFIND is used.

Reply:
Understood!  Changed here and elsewhere where the same incorrect pattern was used.